File: programmers/ForKaKAO/2020/level2/문자열 압축.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string s) {
    int answer = s.length();
    
    for (int i = 1; i <= s.length()/2; i++) {
        int idx = 0;
        int len = s.length();
        
        while (true) {
            int cnt = 0;
            string str = s.substr(idx, i);
            idx += i;
            
            if (idx >= s.length()) {
                break;
            }
            
            while (true) {
                if (str.compare(s.substr(idx, i)) == 0) {
                    cnt++;
                    idx += i;
                } else {
                    break;
                }
            }
            if (cnt > 0) {
                len -= i * cnt;
                
                if (cnt < 9) {
                    len += 1;
                } else if (cnt < 99) {
                    len += 2;
                } else if (cnt < 999) {
                    len += 3;
                } else {
                    len += 4;
                }
            }
            answer = min(answer, len);
        }
    }
    return answer;
}

File: programmers/ForKaKAO/2020/level2/문자열 압축.py
def solution(s):
    answer = len(s)
    
    for i in range(1, int(len(s)/2) + 1):
        pos = 0
        compress = len(s)
        
        while pos + i <= len(s):
            unit = s[pos:pos + i]
            pos += i
            cnt = 0
            
            while pos + i <= len(s):
                if unit == s[pos:pos + i]:
                    cnt += 1
                    pos += i
                else:
                    break
            
            if cnt > 0:
                compress -= i * cnt
                
                if cnt < 9:
                    compress += 1
                elif cnt < 99:
                    compress += 2
                elif cnt < 999:
                    compress += 3
                else:
                    compress += 4
                    
        answer = min(answer, compress)
                    
    return answer